Add Kotlin language, with a little extra time to run.

SirBogman opened the pull request:
See issue #151

This mostly works, but not fully. Rock-paper-scissors-Spock-lizard doesn't seem to run. Perhaps it has something to do with the arguments. I didn't test all holes.

SirBogman commented:
Also the image is large and the alpine is probably old but I don't have the time right now to address.

JRaspass commented on langs/kotlin/Dockerfile:
```diff
@@ -0,0 +1,36 @@
+
+ENTRYPOINT ["/bin/bash"]
+
+FROM scratch
```
`FROM codegolf/lang-base` would remove a lot of the `/empty` boilerplate.

JRaspass commented on langs/kotlin/Dockerfile:
```diff
@@ -0,0 +1,36 @@
+RUN mkdir /myetc        \
+ && echo nobody:x:65534:99::/: > /myetc/passwd
+
+RUN apk add --no-cache curl zip unzip bash strace build-base linux-headers
```
`strace` can go.

JRaspass commented on sql/a-schema.sql:
```diff
@@ -67,7 +67,7 @@ CREATE TYPE lang AS ENUM (
     'elixir', 'f-sharp', 'factor', 'fish', 'forth', 'fortran', 'go',
     'golfscript', 'haskell', 'hexagony', 'j', 'janet', 'java', 'javascript',
-    'jq', 'julia', 'k', 'lisp', 'lua', 'nim', 'ocaml', 'pascal', 'perl',
+    'jq', 'julia', 'k', 'kotlin', 'lisp', 'lua', 'nim', 'ocaml', 'pascal', 'perl',
```
Experimental langs shouldn't update the enum.

JRaspass commented on langs/kotlin/Dockerfile:
```diff
@@ -0,0 +1,36 @@
+FROM adoptopenjdk/openjdk11:x86_64-alpine-jdk-11.0.12_7-slim as builder
```
Maybe use the same base image as Java & Rockstar?

Yewzir commented:
@SirBogman

If it's even the slightest help for dependencies whatsoever, I'm happy to share the Dockerfile I had constructed up until last week or so. 😃 The resulting image size was 272 MB, but I wasn't looking forward to filtering Kotlin's library files.

```Dockerfile
FROM eclipse-temurin:23_37-jdk-alpine AS builder

RUN apk add --no-cache bash curl

ENV VER=2.0.21

RUN curl -#Lo kotlin.zip https://github.com/JetBrains/kotlin/releases/download/v$VER/kotlin-compiler-$VER.zip

RUN unzip kotlin.zip -d / \
 && cp -r /kotlinc/* /usr

FROM codegolf/lang-base

COPY --from=0 /bin/bash /bin/uname                          /bin/
COPY --from=0 /lib/ld-musl-x86_64.so.1                      \
              /lib/libacl.so.1                              \
              /lib/libattr.so.1                             \
              /lib/libcrypto.so.3                           \
              /lib/libskarnet.so.2.14                       \
              /lib/libutmps.so.0.1                          \
              /lib/libz.so.1                                /lib/
COPY --from=0 /opt/java/openjdk/bin/java                    /opt/jdk/bin/
COPY --from=0 /opt/java/openjdk/conf/security/java.security /opt/jdk/conf/security/
COPY --from=0 /opt/java/openjdk/lib/jrt-fs.jar              \
              /opt/java/openjdk/lib/jvm.cfg                 \
              /opt/java/openjdk/lib/libjava.so              \
              /opt/java/openjdk/lib/libjimage.so            \
              /opt/java/openjdk/lib/libjli.so               \
              /opt/java/openjdk/lib/libnet.so               \
              /opt/java/openjdk/lib/libnio.so               \
              /opt/java/openjdk/lib/libverify.so            \
              /opt/java/openjdk/lib/libzip.so               \
              /opt/java/openjdk/lib/modules                 /opt/jdk/lib/
COPY --from=0 /opt/java/openjdk/lib/server/libjvm.so        /opt/jdk/lib/server/
COPY --from=0 /usr/build.txt                                /usr/
COPY --from=0 /usr/bin/dirname                              \
              /usr/bin/env                                  \
              /usr/bin/kotlin                               \
              /usr/bin/kotlinc                              /usr/bin/
COPY --from=0 /usr/lib/*.jar                                \
              /usr/lib/libncursesw.so.6                     \
              /usr/lib/libreadline.so.8                     /usr/lib/

ENV JAVA_HOME=/opt/jdk

ENTRYPOINT ["kotlin"]

CMD ["-version"]
```

I've never really looked into the approaches that involve C, and some seem very interesting. Code evaluation with Kotlin allows getting rid of the entry point definition, though. 😃 Unfortunately, I'm aware of the fact that it doesn't allow for much tweaking. This was my `config/data/langs.toml` entry.

```toml
[Kotlin]
args       = [ '/usr/bin/kotlin', '-e', '$code' ]
env        = [ 'JAVA_HOME=/opt/jdk' ]
experiment = 151
size       = '645 MiB'
version    = '2.0.21'
website    = 'https://kotlinlang.org'
example    = '''
// Printing
println("Hello, World!")

// Looping
for (i in 0..9)
    println(i)

// Accessing arguments
for (arg in args)
    println(arg)
'''
```
Well done, pal! 😄

SirBogman commented:
@Yewzir thanks, I'll have to take a look later.

I have addressed the feedback and fixed the timeout extension. It turns out there wasn't a problem in Rock-paper-scissors-Spock-lizard , just with the timeout extension.
The docker image is way too big, but lets see how it performs on the real server.
